pkg/couchdb/cursor.go

```diff
@@ -1,19 +1,96 @@
 package couchdb
 
-// Cursor holds a pointer in a couchdb map reduce results
-type Cursor struct {
-	Limit     int
-	Done      bool
+// A Cursor holds a reference to a page in a couchdb View
+type Cursor interface {
+	HasMore() bool
+	ApplyTo(req *ViewRequest) *ViewRequest
+	UpdateFrom(res *ViewResponse)
+}
+
+// NewKeyCursor returns a new key based Cursor pointing to
+// the given start_key & startkey_docid
+func NewKeyCursor(limit int, key interface{}, id string) Cursor {
+	return &StartKeyCursor{
+		baseCursor: &baseCursor{Limit: limit},
+		NextKey:    key,
+		NextDocID:  id,
+	}
+}
+
+// NewSkipCursor returns a new skip based Cursor pointing to
+// the page after skip items
+func NewSkipCursor(limit, skip int) Cursor {
+	return &SkipCursor{
+		baseCursor: &baseCursor{Limit: limit},
+		Skip:       skip,
+	}
+}
+
+type baseCursor struct {
+	// Done will be true if there is no more result after last fetch
+	Done bool
+
+	// Limit is maximum number of items retrieved from a request
+	Limit int
+}
+
+// HasMore returns true if there is more document after the current batch.
+// This value is meaning full only after UpdateFrom
+func (c *baseCursor) HasMore() bool { return !c.Done }
+func (c *baseCursor) updateFrom(res *ViewResponse) {
+	lrows := len(res.Rows)
+	if lrows <= c.Limit {
+		c.Done = true
+	} else {
+		res.Rows = res.Rows[:lrows-1]
+		c.Done = false
+	}
+}
+
+// SkipCursor is a Cursor using Skip to know how deep in the request it is.
+type SkipCursor struct {
+	*baseCursor
+	// Skip is the number of elements to start from
+	Skip int
+}
+
+// ApplyTo applies the cursor to a ViewRequest
+// the transformed ViewRequest will retrieve elements from Cursor to
+// Limit or EndKey whichever comes first
+// /!\ Mutates req
```

Comment:
Maybe making the `ApplyTo` method not return a `ViewRequest` would make it clearer that it is a "mutating" method ?